### DISCUSSION TITLE
Documentation: resolver should return a Promise

### DISCUSSION THREAD

**Discussion by Satyam** [react-hook-form/react-hook-form#2295] [General]
In [API](https://react-hook-form.com/api), the description for the `resolver` says that it should be a function returning an object with `values` and `errors` properties but it actually expects a Promise returning such an object. 

If a resolver returns an object, the `then` at [validateResolver](https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L569) fails.

I don't know what the intent was but my guess is that it should either:
* be documented that `resolver` should return a Promise
* fix [validateResolver](https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L569) 
  * it should use an async function or wrap the resolver in `Promise.resolve` so it can return either.
  * document that a resolver can return either.
  * fix the Resolver type.

**Comment by bluebill1049** [comment]
thanks, you are right will sort out the doc.

**Reply by Satyam** [reply]
Something also missing in the description of the `resolver` function is what is the purpose of the returned `values` object.  It should state, somewhere in there, that the purpose of the resolver is to validate **and to cast or transform the values to be submitted**.

**Comment by bluebill1049** [comment]
feel free to send us a pr next time too.

**Comment by Satyam** [comment]
I forgot to mention that this is a change from earlier pre-V6 versions.  I had a test where I had a mock function just to check whether the validation was being called, and it worked in earlier versions:

```js
      const validate = jest.fn((values) =>
        ({
          values,
          errors: {},
        })
      );
```
Suddenly, it stopped working because of the `then` pointed out above.  So, there is a backward compatibility issue at stake.

So, I'd love to make a PR, but which one?

**Reply by bluebill1049** [reply]
can you change to async func see if that will fix it? otherwise provide a codesandbox, happy to take a look at it.

**Reply by Satyam** [reply]
Done: https://github.com/react-hook-form/react-hook-form/pull/2305
